File: python/src/ancilis/evidence/adapter.py
"""Plugin evidence adapter contracts and explicit runtime selection."""

from __future__ import annotations

import logging
from collections.abc import Mapping
from dataclasses import dataclass, field
from types import MappingProxyType
from typing import Any, Protocol, runtime_checkable

from ancilis.config import ResolvedConfig
from ancilis.evidence.record import EvidenceRecord
from ancilis.plugins import PluginContext, PluginRegistry

logger = logging.getLogger(__name__)


def _empty_mapping() -> Mapping[str, Any]:
    return MappingProxyType({})


@dataclass(frozen=True)
class EvidenceAdapterPayload:
    """Canonical SDK evidence plus adapter-local metadata."""

    record: EvidenceRecord
    adapter_metadata: Mapping[str, Any] = field(default_factory=_empty_mapping)


@dataclass(frozen=True)
class EvidenceAdapterQuery:
    """Optional evidence query filters exposed to plugin adapters."""

    agent_id: str | None = None
    session_id: str | None = None
    tool_name: str | None = None
    decision: str | None = None
    since: str | None = None
    limit: int | None = 100


@dataclass(frozen=True)
class EvidenceAdapterExport:
    """Evidence export request exposed to plugin adapters."""

    format: str = "json"
    query: EvidenceAdapterQuery = field(default_factory=EvidenceAdapterQuery)


@runtime_checkable
class EvidenceAdapter(Protocol):
    """Plugin evidence adapter shape supported by the SDK runtime."""

    def store(self, payload: EvidenceAdapterPayload) -> None:
        """Store or forward a canonical evidence payload."""
        ...

    def query(self, query: EvidenceAdapterQuery | None = None) -> object:
        """Query adapter-managed evidence."""
        ...

    def export(self, export: EvidenceAdapterExport | None = None) -> object:
        """Export adapter-managed evidence."""
        ...


@dataclass(frozen=True)
class EvidenceAdapterSelection:
    """Resolved plugin evidence adapter plus non-fatal selection warnings."""

    adapter: EvidenceAdapter | None
    warnings: tuple[str, ...] = ()


def resolve_evidence_adapter(
    config: ResolvedConfig,
    *,
    plugin_name: str | None = None,
    plugin_configs: Mapping[str, Mapping[str, Any]] | None = None,
    plugin_registry: PluginRegistry | None = None,
) -> EvidenceAdapterSelection:
    """Resolve one explicitly selected plugin evidence adapter.

    No adapter is selected by default, leaving the DuckDB-backed ``EvidenceStore``
    as the SDK-owned canonical evidence store.
    """

    del config
    warnings: list[str] = []
    adapter_name = _requested_adapter_name(plugin_name, warnings)
    if adapter_name is None:
        return EvidenceAdapterSelection(adapter=None, warnings=tuple(warnings))

    discovered = plugin_registry or PluginRegistry.discover()
    record = next(
        (
            candidate
            for candidate in discovered.compatible("adapter")
            if candidate.name == adapter_name and candidate.metadata is not None
        ),
        None,
    )
    if record is None:
        _warn(warnings, f"Plugin evidence adapter '{adapter_name}' was not discovered or compatible.")
        return EvidenceAdapterSelection(adapter=None, warnings=tuple(warnings))
    if record.plugin is None:
        _warn(warnings, f"Plugin evidence adapter '{adapter_name}' has no plugin object and was skipped.")
        return EvidenceAdapterSelection(adapter=None, warnings=tuple(warnings))

    plugin_config_map = plugin_configs or {}
    context = PluginContext(
        sdk_version=discovered.sdk_version,
        config=plugin_config_map.get(adapter_name, {}),
    )
    try:
        adapter = record.plugin.create_adapter(context)  # type: ignore[attr-defined]
    except Exception as exc:
        _warn(warnings, f"failed to create plugin evidence adapter '{adapter_name}': {exc}")
        return EvidenceAdapterSelection(adapter=None, warnings=tuple(warnings))

    if not isinstance(adapter, EvidenceAdapter):
        _warn(
            warnings,
            f"Plugin evidence adapter '{adapter_name}' did not expose store(), query(), and export().",
        )
        return EvidenceAdapterSelection(adapter=None, warnings=tuple(warnings))

    return EvidenceAdapterSelection(adapter=adapter, warnings=tuple(warnings))


def _requested_adapter_name(plugin_name: str | None, warnings: list[str]) -> str | None:
    if plugin_name is None:
        return None
    if not plugin_name.startswith("plugin:"):
        _warn(
            warnings,
            f"Plugin evidence adapter selector '{plugin_name}' is not namespaced as plugin:<name>.",
        )
        return None
    name = plugin_name.split(":", 1)[1].strip()
    if not name:
        _warn(warnings, "Empty plugin evidence adapter selector was skipped.")
        return None
    return name


def _warn(warnings: list[str], message: str) -> None:
    warnings.append(message)
    logger.warning(message)


__all__ = [
    "EvidenceAdapter",
    "EvidenceAdapterExport",
    "EvidenceAdapterPayload",
    "EvidenceAdapterQuery",
    "EvidenceAdapterSelection",
    "resolve_evidence_adapter",
]

File: typescript/src/ancilis/evidence/adapter.ts
/** Plugin evidence adapter contracts and explicit runtime selection. */

import type { ResolvedConfig } from "../config/index.js";
import type { EvidenceRecord } from "./record.js";
import { PluginRegistry, type PluginContext, type PluginRecord } from "../plugins/index.js";

export interface EvidenceAdapterPayload {
  readonly record: EvidenceRecord;
  readonly adapterMetadata: Readonly<Record<string, unknown>>;
}

export interface EvidenceAdapterQuery {
  readonly agentId?: string | null;
  readonly sessionId?: string | null;
  readonly toolName?: string | null;
  readonly decision?: string | null;
  readonly since?: string | null;
  readonly limit?: number | null;
}

export interface EvidenceAdapterExport {
  readonly format?: string;
  readonly query?: EvidenceAdapterQuery;
}

export interface EvidenceAdapter {
  store(payload: EvidenceAdapterPayload): unknown;
  query(query?: EvidenceAdapterQuery): unknown;
  export(exportRequest?: EvidenceAdapterExport): unknown;
}

export interface EvidenceAdapterSelection {
  readonly adapter: EvidenceAdapter | null;
  readonly warnings: readonly string[];
}

export interface ResolveEvidenceAdapterOptions {
  readonly pluginName?: string | null;
  readonly pluginConfigs?: Readonly<Record<string, Readonly<Record<string, unknown>>>>;
  readonly pluginRegistry?: PluginRegistry;
  readonly rootDir?: string;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isEvidenceAdapter(value: unknown): value is EvidenceAdapter {
  return (
    isRecord(value)
    && typeof value.store === "function"
    && typeof value.query === "function"
    && typeof value.export === "function"
  );
}

function warn(warnings: string[], message: string): void {
  warnings.push(message);
}

function requestedAdapterName(pluginName: string | null | undefined, warnings: string[]): string | null {
  if (pluginName === undefined || pluginName === null) return null;
  if (!pluginName.startsWith("plugin:")) {
    warn(warnings, `Plugin evidence adapter selector '${pluginName}' is not namespaced as plugin:<name>.`);
    return null;
  }
  const name = pluginName.split(":", 2)[1]?.trim() ?? "";
  if (name.length === 0) {
    warn(warnings, "Empty plugin evidence adapter selector was skipped.");
    return null;
  }
  return name;
}

async function loadAdapterPlugin(
  registry: PluginRegistry,
  record: PluginRecord,
): Promise<unknown> {
  return record.plugin ?? registry.load(record);
}

export async function resolveEvidenceAdapter(
  config: ResolvedConfig,
  options: ResolveEvidenceAdapterOptions = {},
): Promise<EvidenceAdapterSelection> {
  void config;
  const warnings: string[] = [];
  const adapterName = requestedAdapterName(options.pluginName, warnings);
  if (adapterName === null) {
    return { adapter: null, warnings };
  }

  const pluginRegistry = options.pluginRegistry ?? await PluginRegistry.discover({ rootDir: options.rootDir });
  const record = pluginRegistry
    .compatible("adapter")
    .find((candidate) => candidate.name === adapterName && candidate.metadata !== null);
  if (!record) {
    warn(warnings, `Plugin evidence adapter '${adapterName}' was not discovered or compatible.`);
    return { adapter: null, warnings };
  }

  let loadedPlugin: unknown;
  try {
    loadedPlugin = await loadAdapterPlugin(pluginRegistry, record);
  } catch {
    warn(warnings, `Plugin evidence adapter '${adapterName}' was not discovered or compatible.`);
    return { adapter: null, warnings };
  }

  if (!isRecord(loadedPlugin) || typeof loadedPlugin.createAdapter !== "function") {
    warn(warnings, `Plugin evidence adapter '${adapterName}' did not expose store(), query(), and export().`);
    return { adapter: null, warnings };
  }

  const context: PluginContext = {
    sdkVersion: pluginRegistry.sdkVersion,
    config: options.pluginConfigs?.[adapterName] ?? {},
  };

  let adapter: unknown;
  try {
    adapter = loadedPlugin.createAdapter(context);
  } catch (error: unknown) {
    warn(warnings, `failed to create plugin evidence adapter '${adapterName}': ${(error as Error).message ?? String(error)}`);
    return { adapter: null, warnings };
  }

  if (!isEvidenceAdapter(adapter)) {
    warn(warnings, `Plugin evidence adapter '${adapterName}' did not expose store(), query(), and export().`);
    return { adapter: null, warnings };
  }

  return { adapter, warnings };
}
